package/MDAnalysis/topology/tpr/utils.py

```diff
@@ -157,57 +146,25 @@ def from_unpacker(cls, unpacker):
                 raise ValueError("Unrecognized precision")
         return new_unpacker
 
-    def _unpack_value(self, item_size, struct_template):
-        padded_size = self.base_size * item_size
-        start_position = self._pos
-        end_position = self._pos = start_position + padded_size
-        full_content = self._buf[start_position:end_position]
-        if len(full_content) != padded_size:
-            raise EOFError
-        content = full_content[self.base_size - 1::self.base_size]
-        return struct.unpack(struct_template, content)[0]
-
-    def unpack_float(self):
-        return self._unpack_value(4, '<f')
-
-    def unpack_double(self):
-        return self._unpack_value(8, '<d')
-
-    def unpack_int(self):
-        return self._unpack_value(4, '<i')
-
-    def unpack_uint(self):
-        return self._unpack_value(4, '<I')
-
-    def unpack_ushort(self):
-        return self._unpack_value(2, '<H')
-
-    def unpack_char(self):
-        return self._unpack_value(1, '<c')
-
-    def unpack_uchar(self):
-        return self._unpack_value(1, '>B')
-
-    def unpack_string(self):
-        self._pos -= 4 * self.base_size
-        n = self.unpack_int()
-        self.unpack_uint()
-        value = self.unpack_fstring(n)
-        return value
-
     def unpack_fstring(self, n):
         if n < 0:
             raise ValueError('Size of fstring cannot be negative.')
         start_position = self._pos
-        end_position = self._pos = start_position + n * self.base_size
+        end_position = self._pos = start_position + n
         if end_position > len(self._buf):
             raise EOFError
-        full_content = self._buf[start_position:end_position]
-        content = full_content[self.base_size - 1::self.base_size]
+        content = self._buf[start_position:end_position]
         return content
 
-    def unpack_farray(self, n, unpack_item):
-        return [unpack_item() for _ in range(n)]
+    def unpack_ushort(self):
+        # The InMemorySerializer implements ushort according to the XDR standard
+        # on the contrary to the IO serializer.
+        return self._unpack_value(2, '>H')
+
+    def unpack_uchar(self):
+        # The InMemorySerializer implements uchar according to the XDR standard
+        # on the contrary to the IO serializer.
+        return self._unpack_value(1, '>B')
 
 
 def do_string(data):
```

Comment:
I guess that `do_string` has to be different for the 2020 files as well: The `InMemorySerializer` [packs the length as `uint64`](https://github.com/gromacs/gromacs/blob/6edce54dff1102ed59b8ab1c671b18c1702640bd/src/gromacs/utility/inmemoryserializer.cpp#L113-L117). The current implementation discards the first 32 bits (by reading an `int`) and then [tries to interpret the next 32 bits as `unsigned int`](https://github.com/python/cpython/blob/master/Lib/xdrlib.py#L215-L217).

Reply:
The IO serializer in gromacs used to do a few odd things that are now fixed or add in a different way...